src/en/charms-offline-strategies.md

```diff
@@ -0,0 +1,145 @@
+
+In general, all instances within a cloud should be able to communicate directly
+with one another. It may, or it may not, be necessary to express this via the
+`no-proxy` variable. This is because the latter is limited to the HTTP and
+HTTPS protocols, which may not apply in every case. Nevertheless, it is often
+done out of simplicity.
+
+#### No proxy and the localhost cloud
+
+This "no proxy" idea is often used extensively with the localhost cloud at both
+the (shell) environment level and at the Juju level. The purpose being to
+ensure the client can connect with the local controller and that the local
+machines can inter-communicate, all without going through a proxy.
+
+Hosts to exclude from the proxy include:
+
+ - localhost & 127.0.0.1 (standard 'no proxy' settings)
+ - the address assigned to the regular Ethernet device
+ - the address assigned to the host's LXD bridge (to talk to the controller via
+   its API)
+ - the entire subnet dedicated to the particular LXD installation
+
+## APT proxy
+
+An HTTP/S or FTP proxy may not be configured to accept requests for APT
+packages. An APT proxy is identical to that of an HTTP/S or FTP proxy except
+that it applies specifically to APT package requests.
+
+Some common implementations include:
+
+ - [`APT-cacher`][upstream-apt-cacher]
+ - [`Apt-Cacher NG`][upstream-apt-cacher-ng]
+ - [`squid`][upstream-squid]
+ - [`squid-deb-proxy`][upstream-squid-deb-proxy] (based on squid)
+
+Juju uses options `apt-ftp-proxy`, `apt-http-proxy`, and `apt-https-proxy` to
+set these proxies, depending on the protocol involved (i.e. FTP, HTTP, or
+HTTPS).
+
+## FTP proxy
+
+A standard FTP proxy. Juju uses the `ftp-proxy` option for this type of proxy.
+
+## APT mirror
+
+Instead of proxying client requests to an internet-based repository it is
+possible to maintain the repository internally. That is, you can have a copy or
+*mirror* of a Ubuntu package repository. This option has a large storage
+requirement and the initial setup/download time is considerable. Regular mirror
+synchronization will also be needed.
+
+Here are some popular mirroring solutions:
+
+ - [`apt-mirror`][upstream-apt-mirror]
+ - [`debmirror`][upstream-debmirror]
+ - [`aptly`][upstream-aptly]
+
+Juju uses the `apt-mirror` option for this.
+
+### HTTP/S server
+
+For any sort of mirror, an HTTP/S server (i.e. a web server) will be required
+to respond to the actual client requests. These are the most common ones:
+
+ - [`Apache`][upstream-apache]
+ - [`nginx`][upstream-nginx]
+ - [`lighttpd`][upstream-lighttpd]
+
+## Juju agent mirror
+
+An agent that gets installed onto a Juju machine is proxied through the
+controller. If the latter cannot satisfy the request via its cache it will
+download the agent from the [official agent site][upstream-agents], and then
+pass it on to the machine. It is possible, however, to set up an agent mirror
+so the remote site is not solicited by the controller. Download the latest few
```

Comment:
How do we download agents?

Reply:
Someone who is reading about offline deployments, which is an advanced topic, should be able to figure out how to download a [file](https://streams.canonical.com/juju/tools/agent/2.3.1/) from a given URL. I'm open to be corrected on this but I'm mainly concerned about consistency of style with the rest of the page. For example, I also do not show how to set up a web server. Maybe I can list an example file (e.g. `juju-2.3.1-ubuntu-amd64.tgz`)?

Reply:
But you aren't telling them to download a file, you just mention a website in passing. As it turns out, the sync-agent-binaries command can be used for this if you want specific versions.
I realise this is about 'strategies', but if you don't give people enough information to make the link to something practical, it isn't quite so useful. There are 10 thousand ways to set up a website and a million sources online to show you how to do it, but there aren't any practical guides to managing juju agent binaries offline that I am aware of.

Reply:
Yes, the `sync-agent-binaries` command can be used to download agents locally. I will mention it.

Reply:
Done